### ISSUE TITLE
[Feature Request]: adopt xdg-desktop-portal GlobalShortcuts API

### ISSUE THREAD

**Issue by jokeyrhyme** [electron/electron#38288]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

- modern Linux environments such as wayland and flatpak prevent applications from direct access to user input events (for security purposes)
- traditional approaches for implementing global keyboard shortcuts don't work here

### Proposed Solution

- there's a new GlobalShortcuts portal API: https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.GlobalShortcuts
- when an electron consumer calls `globalShortcut` functions ( https://www.electronjs.org/docs/latest/api/global-shortcut ) they should transparently attempt to interact with xdg-desktop-portal via DBus (when available)
- this way, app developers do not need to make any other changes to their code (besides updating the version of electron they use) 

### Alternatives Considered

- it might be worth making this a new JS API instead, but I predict that would lead to poor adoption and thus increased frustration for Linux users of electron-based apps
- other approaches include using root permissions to bypass the wayland / flatpak security model in order to maintain direct access to user input events, but this seems a lot less safe than the proposed solution

### Additional Information

These Linux desktop environments have already implemented the GlobalShortcuts portal:
- Hyprland: https://github.com/hyprwm/xdg-desktop-portal-hyprland/blob/2bdadd2d37f42b516587c64c963ea638c6df8355/hyprland.portal#L3
- KDE: https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/blob/faf620bae9a6601e5380de96ceebce63bc7a062d/data/kde.portal#L3

These are tracking issues/discussions for other desktop environments:
- COSMIC: https://github.com/pop-os/xdg-desktop-portal-cosmic/issues/4
- GNOME: https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/issues/47
- wlroots (for sway/river and others): https://github.com/emersion/xdg-desktop-portal-wlr/issues/240

**Comment by Nantris** [comment]
It seems like this would address #42573 and #15863.

Wayland finally has practically all blockers resolved after many years, and as a result it is increasingly popular. This, or a comparable solution, really must be implemented for Electron to be living up to what it promises. Entire applications rely on the `globalShortcut` feature.

Users are going to be forced onto Wayland in the coming months and there isn't even a meaningful discussion yet on this topic. That discussion really needs to begin ASAP.

**Comment by Nantris** [comment]
Friendly bump. The window to address this without massive disruptions is closing.

**Comment by codebytere** [comment]
I understand this as a goal but we're not prepared to or interested in deviating from Chromium in order to accomplish it. Our entire `globalShortcuts` mechanism is directly plumbed to `chrome/browser/extensions/global_shortcut_listener.h`.

Your best path forward here is to raise an issue with Chromium at https://issues.chromium.org/issues and ask them to update their underlying implementation.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/electron/electron/issues/38288#issuecomment-2241026846

If anyone creates a Chromium issue, link it here. Otherwise, if nobody has, I'll do so.

**Comment by benstigsen** [comment]
> > [#38288 (comment)](https://github.com/electron/electron/issues/38288#issuecomment-2241026846)
> 
> If anyone creates a Chromium issue, link it here. Otherwise, if nobody has, I'll do so.

As far as I can tell, there is already an open issue about this (https://issues.chromium.org/issues/40759171)

Quote:
> My understanding is this is still blocked on [crbug.com/1196508](http://crbug.com/1196508), because whether the interface is actually needed or not is not clear.
Note that, if we need this for wayland, at least we need extension API (i.e. we cannot implement on top of the current protocol fully, AFAIK).
If needed, we can add that to exo (on CrOS), but adding it to Linux (e.g. weston) is another bar.

**Comment by lifebound** [comment]
This looks to have been merged Nov 19 2024 https://chromium-review.googlesource.com/c/chromium/src/+/5871484

**Comment by benstigsen** [comment]
> [@codebytere](https://github.com/codebytere) can we get this re-opened in light of the news above? Or otherwise should we create a new issue? Lack of global shortcuts is killing us.

Electron is based on Chromium. So `xdg-desktop-portal` will be included when they make use of the Chromium version included in that patch. Re-opening does not make sense, as it has now been solved, at this point we're just waiting.

**Comment by Nantris** [comment]
@benstigsen you're right. I'm conflating this with issue #42573.

**Comment by Malix-Labs** [comment]
https://github.com/electron/electron/issues/38288#issuecomment-2492287687 from @benstigsen :
> it has now been solved. Now we just have to wait.

#45607 seems to suggest otherwise (or is it unrelated?), in which case this issue should be reopened

Else, it can stay closed as not planned

**Comment by benstigsen** [comment]
> [#38288 (comment)](https://github.com/electron/electron/issues/38288#issuecomment-2492287687) from [@benstigsen](https://github.com/benstigsen) :
> 
> > it has now been solved. Now we just have to wait.
> 
> [#45607](https://github.com/electron/electron/issues/45607) seems to suggest otherwise (or is it unrelated?), in which case this issue should be reopened
> 
> Else, it can stay closed as not planned

As far as I can tell, that issue is related to a bug of the feature, whereas this issue is about the feature itself. If the bug still exists, the issue you link should be re-opened. Not this one.

**Comment by jokeyrhyme** [comment]
what's the oldest version of electron that has this feature? or is it yet to be released? cheers! <3

**Comment by Nantris** [comment]
It's not clear that global shortcuts work even in the versions that have this change (and I'm not sure which do sorry.) I just wanted to let you know that personally global shortcuts remain dead for us on Wayland (which is a big problem since perhaps our biggest feature is the global shortcut.)
